File: core/mill-metadata-core/src/main/java/io/qpointz/mill/metadata/domain/core/ValueMappingFacet.java
package io.qpointz.mill.metadata.domain.core;

import com.fasterxml.jackson.annotation.JsonProperty;
import io.qpointz.mill.metadata.domain.AbstractFacet;
import io.qpointz.mill.metadata.domain.MetadataFacet;
import io.qpointz.mill.metadata.domain.ValidationException;
import lombok.Data;
import lombok.EqualsAndHashCode;

import java.util.ArrayList;
import java.util.List;

/**
 * Value mapping facet - maps user terms to database values for NL2SQL.
 * Entity-bound facet attached to ATTRIBUTE entities.
 */
@Data
@EqualsAndHashCode(callSuper = true)
public class ValueMappingFacet extends AbstractFacet {
    
    private String context;
    private Double similarityThreshold;
    private List<ValueMapping> mappings = new ArrayList<>();
    private List<ValueMappingSource> sources = new ArrayList<>();
    
    /**
     * Static value mapping entry.
     */
    public record ValueMapping(
        @JsonProperty("user-term") String userTerm,
        @JsonProperty("database-value") String databaseValue,
        @JsonProperty("display-value") String displayValue,
        @JsonProperty("description") String description,
        @JsonProperty("language") String language,
        @JsonProperty("aliases") List<String> aliases
    ) {
        public ValueMapping {
            if (language == null || language.isEmpty()) {
                language = "en";
            }
            if (aliases == null) {
                aliases = new ArrayList<>();
            }
        }
    }
    
    /**
     * Dynamic value mapping source (SQL query, API, etc.).
     */
    public record ValueMappingSource(
        @JsonProperty("type") String type,
        @JsonProperty("name") String name,
        @JsonProperty("definition") String definition,  // YAML uses "definition", can also be "sql"
        @JsonProperty("description") String description,
        @JsonProperty("enabled") Boolean enabled,
        @JsonProperty("cronExpression") String cronExpression,
        @JsonProperty("cache-ttl-seconds") Integer cacheTtlSeconds
    ) {
        public ValueMappingSource {
            if (enabled == null) {
                enabled = true;
            }
            if (cacheTtlSeconds == null) {
                cacheTtlSeconds = 3600;
            }
        }
        
        /**
         * Get SQL definition, checking both "definition" and "sql" fields.
         */
        public String getSql() {
            return definition != null ? definition : null;
        }
        
        /**
         * Check if source is enabled.
         */
        public boolean isEnabled() {
            return enabled != null && enabled;
        }
        
        /**
         * Get cache TTL in seconds.
         */
        public int getCacheTtl() {
            return cacheTtlSeconds != null ? cacheTtlSeconds : 3600;
        }
    }
    
    @Override
    public String getFacetType() {
        return "value-mapping";
    }
    
    @Override
    public void validate() throws ValidationException {
        if (mappings != null) {
            for (ValueMapping mapping : mappings) {
                if (mapping.userTerm() == null || mapping.userTerm().isEmpty()) {
                    throw new ValidationException("ValueMappingFacet: userTerm is required for mapping");
                }
                if (mapping.databaseValue() == null || mapping.databaseValue().isEmpty()) {
                    throw new ValidationException("ValueMappingFacet: databaseValue is required for mapping: " + mapping.userTerm());
                }
            }
        }
        if (sources != null) {
            for (ValueMappingSource source : sources) {
                if (source.name() == null || source.name().isEmpty()) {
                    throw new ValidationException("ValueMappingFacet: name is required for source");
                }
                if (source.getSql() == null || source.getSql().isEmpty()) {
                    throw new ValidationException("ValueMappingFacet: definition/sql is required for source: " + source.name());
                }
            }
        }
    }
    
    @Override
    public MetadataFacet merge(MetadataFacet other) {
        if (!(other instanceof ValueMappingFacet)) {
            return this;
        }
        ValueMappingFacet otherFacet = (ValueMappingFacet) other;
        
        // Merge: other takes precedence for context and threshold
        if (otherFacet.context != null) {
            this.context = otherFacet.context;
        }
        if (otherFacet.similarityThreshold != null) {
            this.similarityThreshold = otherFacet.similarityThreshold;
        }
        
        // Merge mappings: combine lists, avoid duplicates by userTerm + databaseValue
        if (otherFacet.mappings != null && !otherFacet.mappings.isEmpty()) {
            this.mappings = new ArrayList<>(this.mappings);
            for (ValueMapping otherMapping : otherFacet.mappings) {
                boolean exists = this.mappings.stream()
                    .anyMatch(m -> m.userTerm().equals(otherMapping.userTerm()) &&
                                 m.databaseValue().equals(otherMapping.databaseValue()));
                if (!exists) {
                    this.mappings.add(otherMapping);
                }
            }
        }
        
        // Merge sources: combine lists, avoid duplicates by name
        if (otherFacet.sources != null && !otherFacet.sources.isEmpty()) {
            this.sources = new ArrayList<>(this.sources);
            for (ValueMappingSource otherSource : otherFacet.sources) {
                boolean exists = this.sources.stream()
                    .anyMatch(s -> s.name().equals(otherSource.name()));
                if (!exists) {
                    this.sources.add(otherSource);
                }
            }
        }
        
        return this;
    }
}

File: metadata/mill-metadata-core/src/main/kotlin/io/qpointz/mill/metadata/domain/core/ValueMappingFacet.kt
package io.qpointz.mill.metadata.domain.core

import com.fasterxml.jackson.annotation.JsonAlias
import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonProperty
import io.qpointz.mill.metadata.domain.MetadataFacet

/** Normalization mappings between user terms and physical values. */
@JsonIgnoreProperties(ignoreUnknown = true)
data class ValueMappingFacet(
    @JsonProperty("context")
    val context: String? = null,

    @JsonProperty("similarity-threshold")
    val similarityThreshold: Double? = null,

    @JsonProperty("mappings")
    val mappings: List<ValueMapping> = emptyList(),

    @JsonProperty("sources")
    val sources: List<ValueMappingSource> = emptyList(),

    override val facetType: String = "value-mapping",
) : MetadataFacet {

    /** Single term-to-value mapping entry. */
    data class ValueMapping(
        @JsonProperty("user-term") val userTerm: String? = null,
        @JsonProperty("database-value") val databaseValue: String? = null,
        @JsonProperty("display-value") val displayValue: String? = null,
        @JsonProperty("description") val description: String? = null,
        @JsonProperty("language") val language: String = "en",
        @JsonProperty("aliases") val aliases: List<String> = emptyList(),
    )

    /** External source definition that can populate mapping entries. */
    data class ValueMappingSource(
        @JsonProperty("type") val type: String? = null,
        @JsonProperty("name") val name: String? = null,
        @JsonProperty("definition") @JsonAlias("sql") val definition: String? = null,
        @JsonProperty("description") val description: String? = null,
        @JsonProperty("enabled") val enabled: Boolean = true,
        @JsonProperty("cronExpression") val cronExpression: String? = null,
        @JsonProperty("cache-ttl-seconds") val cacheTtlSeconds: Int = 3600,
    ) {
        val sql: String? get() = definition
        val cacheTtl: Int get() = cacheTtlSeconds
    }
}
